src/filter/json_encode_filter.rs

```diff
@@ -0,0 +1,268 @@
+    pub fn new(config: &JSONEncodeFilterConfig) -> JSONEncodeFilter {
+        JSONEncodeFilter {
+            parse_line: config.parse_line,
+        }
+    }
+}
+
+impl filter::Filter for JSONEncodeFilter {
+    fn process(
+        &mut self,
+        event: metric::Event,
+        res: &mut Vec<metric::Event>,
+    ) -> Result<(), filter::FilterError> {
+        match event {
+            metric::Event::Log(l) => if let Some(ref log) = *l {
+                let naive_time = NaiveDateTime::from_timestamp(log.time, 0);
+                let utc_time: DateTime<Utc> = DateTime::from_utc(naive_time, Utc);
+                let metadata = json!({
+                    "time": utc_time.to_rfc3339(),
+                    "path": log.path.clone(),
+                    "tags": log.tags,
+                });
+                let metadata = metadata.as_object().unwrap();
+                // If parse_line is true, and line is parsable as a JSON object, parse it.
+                // Otherwise get an object containing the original line.
+                let value = (if self.parse_line { Some(()) } else { None })
+                    .and_then(|_| serde_json::from_str::<Value>(&log.value).ok())
+                    .and_then(|v| if let Value::Object(obj) = v { Some(obj) } else { None })
+                    .map(|v| { JSON_ENCODE_LOG_PARSED.fetch_add(1, Ordering::Relaxed); v })
+                    .unwrap_or_else(||
+                        if let Value::Object(obj) = json!({"message": log.value.clone()}) {
+                            obj
+                        } else {
+                            unreachable!()
+                        }
+                    );
+                // Combine our various sources of data.
+                // Data that is more likely to be correct (more specific to the source) overrides
+                // other data. So the parsed value is authoritative, followed by any fields we could
+                // parse by filters, then finally the metadata we were able to work out on our own.
+                let value = merge_objects(&[&value, &log.fields.clone().into(), &metadata]);
+                res.push(metric::Event::Raw {
+                    order_by: random(),
+                    encoding: metric::Encoding::JSON,
+                    bytes: serde_json::to_string(&value).unwrap().into(), // serde_json::Value will never fail to encode
+                });
+                JSON_ENCODE_LOG_PROCESSED.fetch_add(1, Ordering::Relaxed);
+            },
+            // All other event types are passed through.
+            event => {
+                res.push(event);
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Merge JSON objects, with values from earlier objects in the list overriding later ones.
+/// Note this is not a recursive merge - if the same key is in many objects, we simply take
+/// the value from the earliest one.
+fn merge_objects(objs: &[&Map<String, Value>]) -> Map<String, Value> {
```

Comment:
You're passing `objs` into this function but then the passed values are never re-used. Consider moving `objs` into this function so you can avoid cloning on line 129.